File: api/src/spp_extras_api/queries/characters.py
from django.db import connections
from spp_extras_api.models.classiccharacters import (
    ClassicCharacterQueststatus,
    ClassicCharacterQueststatusWeekly,
    ClassicCharacterReputation,
    ClassicCharacters
)
from spp_extras_api.models.tbccharacters import (
    TbcCharacterQueststatus,
    TbcCharacterQueststatusDaily,
    TbcCharacterQueststatusMonthly,
    TbcCharacterQueststatusWeekly,
    TbcCharacterReputation,
    TbcCharacters
)
from spp_extras_api.models.wotlkcharacters import (
    WotlkCharacterAchievement,
    WotlkCharacterAchievementProgress,
    WotlkCharacterAchievementSharedProgress,
    WotlkCharacterQueststatus,
    WotlkCharacterQueststatusDaily,
    WotlkCharacterQueststatusMonthly,
    WotlkCharacterQueststatusWeekly,
    WotlkCharacterReputation,
    WotlkCharacters,
    WotlkCharacterSkills,
    WotlkCharacterSpell,
    WotlkItemInstance,
    WotlkMail,
    WotlkMailItems
)


# Change models depending on expansion

def characters_model(expansion):
    if expansion == 'classic':
        return ClassicCharacters
    elif expansion == 'tbc':
        return TbcCharacters
    else:
        return WotlkCharacters


def character_rep_model(expansion):
    if expansion == 'classic':
        return ClassicCharacterReputation
    elif expansion == 'tbc':
        return TbcCharacterReputation
    else:
        return WotlkCharacterReputation


def regular_quest_model(expansion):
    if expansion == 'classic':
        return ClassicCharacterQueststatus
    elif expansion == 'tbc':
        return TbcCharacterQueststatus
    else:
        return WotlkCharacterQueststatus


def daily_quest_model(expansion):
    if expansion == 'tbc':
        return TbcCharacterQueststatusDaily
    else:
        return WotlkCharacterQueststatusDaily


def weekly_quest_model(expansion):
    if expansion == 'classic':
        return ClassicCharacterQueststatusWeekly
    elif expansion == 'tbc':
        return TbcCharacterQueststatusWeekly
    else:
        return WotlkCharacterQueststatusWeekly


def monthly_quest_model(expansion):
    if expansion == 'tbc':
        return TbcCharacterQueststatusMonthly
    else:
        return WotlkCharacterQueststatusMonthly


# ----------------------------------------------------------------
# Characters
# ----------------------------------------------------------------

def sel_all_chars(expansion, account_ids):
    characters = characters_model(expansion).objects\
        .using(f'{expansion}characters')\
        .all()\
        .filter(account__in=account_ids)
    
    if expansion == 'wotlk':
        return characters.values(
            'guid',
            'account',
            'name',
            'race',
            'gender',
            'class_field',
            'knowntitles')
    else:
        return characters.values(
            'guid',
            'account',
            'name',
            'race',
            'class_field')


# ----------------------------------------------------------------
# Achievement Credit
# ----------------------------------------------------------------

def sel_all_char_achs(char_ids):
    return WotlkCharacterAchievement.objects\
        .using('wotlkcharacters')\
        .all()\
        .filter(guid__in=char_ids)\
        .values()


def ins_char_achs(achievements):
    def create_args(a):
        return WotlkCharacterAchievement(
            guid=a['guid'],
            achievement=a['achievement'],
            date=a['date'])
    args = list(map(create_args, achievements))
    WotlkCharacterAchievement.objects\
        .using('wotlkcharacters')\
        .bulk_create(args, ignore_conflicts=True)


# ----------------------------------------------------------------
# Achievement Progress
# ----------------------------------------------------------------

def sel_all_ach_prog(char_ids):
    return WotlkCharacterAchievementProgress.objects\
        .using('wotlkcharacters')\
        .all()\
        .filter(guid__in=char_ids)\
        .values()


def ins_ach_prog(achievements):
    for ach in achievements:
        guid = ach['guid']
        criteria = ach['criteria']
        counter = ach['counter']
        date = ach['date']
        try:
            WotlkCharacterAchievementProgress.objects\
                .using('wotlkcharacters')\
                .get(
                    guid=guid,
                    criteria=criteria)
        except WotlkCharacterAchievementProgress.DoesNotExist:
            WotlkCharacterAchievementProgress.objects\
                .using('wotlkcharacters')\
                .create(
                    guid=guid,
                    criteria=criteria,
                    counter=counter,
                    date=date)
        else:
            WotlkCharacterAchievementProgress.objects\
                .using('wotlkcharacters')\
                .filter(
                    guid=guid,
                    criteria=criteria)\
                .update(
                    counter=counter,
                    date=date)


# ----------------------------------------------------------------
# Achievement Shared Progress
# ----------------------------------------------------------------

def char_ach_shared_prog_exists():
    wotlk_char_connection = connections['wotlkcharacters']
    cursor = wotlk_char_connection.cursor()
    tables = wotlk_char_connection.introspection.table_names(cursor)
    if 'character_achievement_shared_progress' in tables:
        return True


def create_char_ach_shared_prog():
    with connections['wotlkcharacters'].schema_editor() as schema_editor:
        schema_editor.create_model(WotlkCharacterAchievementSharedProgress)


def sel_all_char_ach_shared_prog(bots):
    progress = WotlkCharacterAchievementSharedProgress.objects\
        .using('wotlkcharacters')\
        .all()\
        
    if bots:
        return progress.values('account', 'criteria', 'counter', 'date')
    else:
        return progress\
            .filter(account=0)\
            .values('account', 'criteria', 'counter', 'date')


def ins_char_ach_shared_prog(achievements):
    for ach in achievements:
        account = ach['account']
        criteria = ach['criteria']
        counter = ach['counter']
        date = ach['date']
        try:
            WotlkCharacterAchievementSharedProgress.objects\
                .using('wotlkcharacters')\
                .get(
                    account=account,
                    criteria=criteria)
        except WotlkCharacterAchievementSharedProgress.DoesNotExist:
            WotlkCharacterAchievementSharedProgress.objects\
                .using('wotlkcharacters')\
                .create(
                    account=account,
                    criteria=criteria,
                    counter=counter,
                    date=date)
        else:
            WotlkCharacterAchievementSharedProgress.objects\
                .using('wotlkcharacters')\
                .filter(
                    account=account,
                    criteria=criteria)\
                .update(
                    counter=counter,
                    date=date)


# ----------------------------------------------------------------
# Achievement Reward Titles
# ----------------------------------------------------------------

def upd_reward_titles(titles):
    for t in titles:
        record = WotlkCharacters.objects\
            .using('wotlkcharacters')\
            .get(guid=t['guid'])
        record.knowntitles = t['knowntitles']
        record.save()


# ----------------------------------------------------------------
# Achievement Reward Items
# ----------------------------------------------------------------

def sel_last_item_inst_id():
    return WotlkItemInstance.objects\
        .using('wotlkcharacters')\
        .all()\
        .values('guid')\
        .last()


def ins_reward_item_instances(instances):
    def create_args(i):
        return WotlkItemInstance(
            guid=i['guid'],
            owner_guid=i['owner_guid'],
            itementry=i['itementry'],
            creatorguid=i['creatorguid'],
            giftcreatorguid=i['giftcreatorguid'],
            count=i['count'],
            duration=i['duration'],
            charges=i['charges'],
            flags=i['flags'],
            enchantments=i['enchantments'],
            randompropertyid=i['randompropertyid'],
            durability=i['durability'],
            playedtime=i['playedtime'],
            text=i['text'])
    args = list(map(create_args, instances))
    WotlkItemInstance.objects\
        .using('wotlkcharacters')\
        .bulk_create(args)


def sel_last_mail_id():
    return WotlkMail.objects\
        .using('wotlkcharacters')\
        .all()\
        .values('id')\
        .last()


def ins_reward_mail(mail):
    def create_args(m):
        return WotlkMail(
            id=m['id'],
            messagetype=m['messagetype'],
            stationery=m['stationery'],
            mailtemplateid=m['mailtemplateid'],
            sender=m['sender'],
            receiver=m['receiver'],
            subject=m['subject'],
            body=m['body'],
            has_items=m['has_items'],
            expire_time=m['expire_time'],
            deliver_time=m['deliver_time'],
            money=m['money'],
            cod=m['cod'],
            checked=m['checked'])
    args = list(map(create_args, mail))
    WotlkMail.objects\
        .using('wotlkcharacters')\
        .bulk_create(args)


def ins_reward_mail_items(items):
    def create_args(i):
        return WotlkMailItems(
            mail_id=i['mail_id'],
            item_guid=i['item_guid'],
            item_template=i['item_template'],
            receiver=i['receiver'])
    args = list(map(create_args, items))
    WotlkMailItems.objects\
        .using('wotlkcharacters')\
        .bulk_create(args)


# ----------------------------------------------------------------
# Quests
# ----------------------------------------------------------------

def sel_all_completed_reg_quests(expansion, char_ids):
    return regular_quest_model(expansion).objects\
        .using(f'{expansion}characters')\
        .all()\
        .filter(guid__in=char_ids, status__exact=1)\
        .values()


def sel_all_completed_daily_quests(expansion, char_ids):
    return daily_quest_model(expansion).objects\
        .using(f'{expansion}characters')\
        .all()\
        .filter(guid__in=char_ids)\
        .values()


def sel_all_completed_weekly_quests(expansion, char_ids):
    return weekly_quest_model(expansion).objects\
        .using(f'{expansion}characters')\
        .all()\
        .filter(guid__in=char_ids)\
        .values()


def sel_all_completed_monthly_quests(expansion, char_ids):
    return monthly_quest_model(expansion).objects\
        .using(f'{expansion}characters')\
        .all()\
        .filter(guid__in=char_ids)\
        .values()


# ----------------------------------------------------------------
# Reputation
# ----------------------------------------------------------------

def sel_all_char_rep(expansion, char_ids):
    return character_rep_model(expansion).objects\
        .using(f'{expansion}characters')\
        .all()\
        .filter(guid__in=char_ids)\
        .values('guid', 'faction', 'standing')


def upd_char_rep(reputations):
    for r in reputations:
        WotlkCharacterReputation.objects\
            .using('wotlkcharacters')\
            .all()\
            .filter(
                guid=r['guid'], faction=r['faction'])\
            .update(standing=r['standing'])


# ----------------------------------------------------------------
# Pets and Mounts
# ----------------------------------------------------------------

def sel_char_riding_skills(char_ids):
    return WotlkCharacterSkills.objects\
        .using('wotlkcharacters')\
        .all()\
        .filter(guid__in=char_ids, skill__exact=762)\
        .values('guid', 'value')


def sel_char_pet_mount_spells(char_ids, spell_ids):
    return WotlkCharacterSpell.objects\
        .using('wotlkcharacters')\
        .all()\
        .filter(guid__in=char_ids, spell__in=spell_ids)\
        .values('guid', 'spell')


def ins_char_pet_mount_spells(spells):
    for spell in spells:
        WotlkCharacterSpell.objects\
            .using('wotlkcharacters')\
            .create(
                guid=spell['guid'],
                spell=spell['spell_id'],
                active=1,
                disabled=0)

File: src/main/db/characters.ts
import { send } from '../../utils';

// ----------------------------------------------------------------
// Characters
// ----------------------------------------------------------------

export const selChars = async (conn: any, xpac: any, accts: any, reply?: any) => { // TEMP ANY
  const values = `
    guid,
    account,
    name,
    race,
    class
    ${xpac === 'wotlk' ? ', gender, knownTitles' : ''}
  `;
  const sql = `
    SELECT ${values} FROM characters
    WHERE account IN (?)
  `;
  try {
    const startMsg = 'Fetching character data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [accts]);
    const successMsg = 'Character data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch character data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

// ----------------------------------------------------------------
// Achievement Credit
// ----------------------------------------------------------------

export const selAllCharAchs = async (conn: any, charIDs: any, reply?: any) => { // TEMP ANY
  const sql = `
    SELECT * FROM character_achievement
    WHERE guid IN (?)
  `;
  try {
    const startMsg = 'Fetching character achievement credit data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [charIDs]);
    const successMsg = 'Failed to fetch character achievement credit data!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch character achievement credit data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const insCharAchs = async (conn: any, achievements: any, reply?: any) => { // TEMP ANY
  const columns = 'guid, achievement, date';
  const values = achievements.map((a: any) => { // TEMP ANY
    return [a.guid, a.achievement, a.date];
  });
  const sql = `INSERT IGNORE INTO character_achievement (${columns}) VALUES ?`;
  try {
    const startMsg = 'Saving new character achievement credit...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [values]);
    const successMsg = 'New character achievement credit successfully saved!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to save new character achievement credit!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

// ----------------------------------------------------------------
// Achievement Progress
// ----------------------------------------------------------------

export const selAllAchProg = async (conn: any, charIDs: any, reply?: any) => { // TEMP ANY
  const sql = `
    SELECT * FROM character_achievement_progress
    WHERE guid IN (?)
  `;
  try {
    const startMsg = 'Fetching character achievement progress data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [charIDs]);
    const successMsg = 'Character achievement progress data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch character achievement progress data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const insUpdAchProg = async (conn: any, achievements: any, reply?: any) => { // TEMP ANY
  const columns = 'guid, criteria, counter, date';
  const values = achievements.map((a: any) => { // TEMP ANY
    return [a.guid, a.criteria, a.counter, a.date];
  });
  const sql = `
    INSERT INTO character_achievement_progress (${columns}) VALUES ?
      ON DUPLICATE KEY UPDATE counter=VALUES(counter), date=VALUES(date)
  `;
  try {
    const startMsg = 'Saving new character achievement progress...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [values]);
    const successMsg = 'New character achievement progress successfully saved!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to save new character achievement progress!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

// ----------------------------------------------------------------
// Achievement Shared Progress
// ----------------------------------------------------------------

export const showSharedProg = async (conn: any, reply?: any) => { // TEMP ANY
  const sql = 'SHOW TABLES LIKE "character_achievement_shared_progress"';
  try {
    const startMsg = 'Fetching shared achievement progress table data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql);
    const successMsg = 'Shared achievement progress table data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch shared achievement progress table data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const createSharedProgTable = async (conn: any, reply?: any) => { // TEMP ANY
  const sql = `
    CREATE TABLE character_achievement_shared_progress (
      account INT NOT NULL,
      criteria INT NOT NULL,
      counter INT DEFAULT 0,
      date BIGINT
    )
  `;
  try {
    const startMsg = 'Creating character_achievement_shared_progress table...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql);
    const successMsg = 'Successfully created character_achievement_shared_progress table!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to create character_achievement_shared_progress table!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const selAllCharAchSharedProg = async (conn: any, accountIDs: boolean, reply?: any) => { // TEMP ANY
  const sql = `
    SELECT * FROM character_achievement_shared_progress
    WHERE guid IN (?)
  `;
  try {
    const startMsg = 'Fetching character achievement shared progress data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [accountIDs]);
    const successMsg = 'Character achievement shared progress data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch character achievement shared progress data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const insUpdCharAchSharedProg = async (conn: any, achievements: any, reply?: any) => { // TEMP ANY
  const columns = 'account, criteria, counter, date';
  const values = achievements.map((a: any) => { // TEMP ANY
    return [a.account, a.criteria, a.counter, a.date];
  });
  const sql = `
    INSERT INTO character_achievement_shared_progress (${columns}) VALUES ?
      ON DUPLICATE KEY UPDATE counter=VALUES(counter), date=VALUES(date)
  `;
  try {
    const startMsg = 'Saving new character achievement shared progress...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [values]);
    const successMsg = 'New character achievement shared progress successfully saved!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to save new character achievement shared progress!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

// ----------------------------------------------------------------
// Achievement Reward Titles
// ----------------------------------------------------------------

export const updRewardTitles = async (conn: any, titles: any, reply?: any) => { // TEMP ANY
  const values = titles.map((t: any) => { // TEMP ANY
    return [t.guid, t.knownTitles];
  });
  const sql = `
    INSERT INTO characters (guid, knownTitles) VALUES ?
      ON DUPLICATE KEY UPDATE knownTitles=VALUES(knownTitles)
  `;
  try {
    const startMsg = 'Updating character titles...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [values]);
    const successMsg = 'Character titles updated!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to update character titles!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

// ----------------------------------------------------------------
// Achievement Reward Items
// ----------------------------------------------------------------

export const selLastItemInstID = async (conn: any, reply?: any) => { // TEMP ANY
  const sql = 'SELECT MAX(guid) FROM item_instance';
  try {
    const startMsg = 'Fetching last item instance ID data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql);
    const successMsg = 'Last item instance ID data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch last item instance ID data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const insRewardItemInstances = async (conn: any, instances: any, reply?: any) => { // TEMP ANY
  const columns = `
    guid,
    owner_guid,
    itemEntry,
    creatorGuid,
    giftCreatorGuid,
    count,
    duration,
    charges,
    flags,
    enchantments,
    randomPropertyId,
    durability,
    playedTime,
    text
  `;
  const values = instances.map((i: any) => { // TEMP ANY
    return [
      i.guid,
      i.owner_guid,
      i.itemEntry,
      i.creatorGuid,
      i.giftCreatorGuid,
      i.count,
      i.duration,
      i.charges,
      i.flags,
      i.enchantments,
      i.randomPropertyId,
      i.durability,
      i.playedTime,
      i.text
    ];
  });
  const sql = `INSERT INTO item_instance (${columns}) VALUES ?`;
  try {
    const startMsg = 'Saving new item instance data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [values]);
    const successMsg = 'New item instance data successfully saved!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to save new item instance data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const selLastMailID = async (conn: any, reply?: any) => { // TEMP ANY
  const sql = 'SELECT MAX(id) FROM mail';
  try {
    const startMsg = 'Fetching last mail ID data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql);
    const successMsg = 'Last mail ID data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch last mail ID data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const insRewardMail = async (conn: any, mail: any, reply?: any) => { // TEMP ANY
  const columns = `
    id,
    messageType,
    stationery,
    mailTemplateId,
    sender,
    receiver,
    subject,
    body,
    has_items,
    expire_time,
    deliver_time,
    money,
    cod,
    checked
  `;
  const values = mail.map((m: any, reply?: any) => { // TEMP ANY
    return [
      m.id,
      m.messageType,
      m.stationery,
      m.mailTemplateId,
      m.sender,
      m.receiver,
      m.subject,
      m.body,
      m.has_items,
      m.expire_time,
      m.deliver_time,
      m.money,
      m.cod,
      m.checked
    ];
  });
  const sql = `INSERT INTO mail (${columns}) VALUES ?`;
  try {
    const startMsg = 'Saving new mail data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [values]);
    const successMsg = 'New mail data successfully saved!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to save new mail data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const insRewardMailItems = async (conn: any, items: any, reply?: any) => { // TEMP ANY
  const columns = 'mail_id, item_guid, item_template, receiver';
  const values = items.map((i: any) => { // TEMP ANY
    return [i.mail_id, i.item_guid, i.item_template, i.receiver];
  });
  const sql = `INSERT INTO mail_items (${columns}) VALUES ?`;
  try {
    const startMsg = 'Saving new mail item data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [values]);
    const successMsg = 'New mail item data successfully saved!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to save new mail item data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

// ----------------------------------------------------------------
// Quests
// ----------------------------------------------------------------

export const selCompletedRegQuests = async (conn: any, charIDs: any, reply?: any) => { // TEMP ANY
  const sql = `
    SELECT * FROM character_queststatus
    WHERE guid IN (?) AND status=1
  `;
  try {
    const startMsg = 'Fetching completed regular quest data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [charIDs]);
    const successMsg = 'Completed regular quest data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch completed regular quest data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const selCompletedDailyQuests = async (conn: any, charIDs: any, reply?: any) => { // TEMP ANY
  const sql = `
    SELECT * FROM character_queststatus_daily
    WHERE guid IN (?)
  `;
  try {
    const startMsg = 'Fetching completed daily quest data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [charIDs]);
    const successMsg = 'Completed daily quest data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch completed daily quest data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const selCompletedWeeklyQuests = async (conn: any, charIDs: any, reply?: any) => { // TEMP ANY
  const sql = `
    SELECT * FROM character_queststatus_weekly
    WHERE guid IN (?)
  `;
  try {
    const startMsg = 'Fetching completed weekly quest data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [charIDs]);
    const successMsg = 'Completed weekly quest data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch completed weekly quest data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const selCompletedMonthlyQuests = async (conn: any, charIDs: any, reply?: any) => { // TEMP ANY
  const sql = `
    SELECT * FROM character_queststatus_monthly
    WHERE guid IN (?)
  `;
  try {
    const startMsg = 'Fetching completed monthly quest data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [charIDs]);
    const successMsg = 'Completed monthly quest data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch completed monthly quest data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

// ----------------------------------------------------------------
// Reputation
// ----------------------------------------------------------------

export const selAllCharRep = async (conn: any, charIDs: any, reply?: any) => { // TEMP ANY
  const sql = `
    SELECT guid, faction, standing FROM character_reputation
    WHERE guid IN (?)
  `;
  try {
    const startMsg = 'Fetching character reputation data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [charIDs]);
    const successMsg = 'Character reputation data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch character reputation data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const updCharRep = async (conn: any, reputations: any, reply?: any) => { // TEMP ANY
  const values = reputations.map((r: any) => { // TEMP ANY
    return [r.guid, r.faction, r.standing, r.flags];
  });
  const sql = `
    INSERT INTO characters (guid, faction, standing, flags) VALUES ?
      ON DUPLICATE KEY UPDATE standing=VALUES(standing)
  `;
  try {
    const startMsg = 'Updating character reputation standings...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [values]);
    const successMsg = 'Character reputation standings updated!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to update character reputation standings!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

// ----------------------------------------------------------------
// Pets and Mounts
// ----------------------------------------------------------------

export const selCharRidingSkills = async (conn: any, charIDs: any, reply?: any) => { // TEMP ANY
  const sql = `
    SELECT guid, value FROM character_skills
    WHERE guid IN (?) AND skill=762
  `;
  try {
    const startMsg = 'Fetching character riding skill data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [charIDs]);
    const successMsg = 'Character riding skill data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch character riding skill data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const selCharPetMountSpells = async (conn: any, charIDs: any, spellIDs: any, reply?: any) => { // TEMP ANY
  const sql = `
    SELECT guid, spell FROM character_spell
    WHERE guid IN (?) AND spell IN (?)
  `;
  try {
    const startMsg = 'Fetching character pet and mount spell data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [charIDs, spellIDs]);
    const successMsg = 'Character pet and mount spell data fetched!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to fetch character pet and mount spell data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}

export const insCharPetMountSpells = async (conn: any, spells: any, reply?: any) => { // TEMP ANY
  const columns = 'guid, spell, active, disabled';
  const values = spells.map((s: any) => { // TEMP ANY
    return [
      s.guid,
      s.spell,
      1, // active
      0 // disabled
    ];
  });
  const sql = `INSERT INTO character_spell (${columns}) VALUES ?`;
  try {
    const startMsg = 'Saving new pet and mount spell data...';
    send(startMsg, reply);
    const [rows] = await conn.query(sql, [values]);
    const successMsg = 'New pet and mount spell data successfully saved!';
    send(successMsg, reply);
    return rows;
  } catch (err) {
    const errMsg = `Failed to save new pet and mount spell data!\n${err}`;
    send(errMsg, reply);
    throw err;
  }
}
